walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,79 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
+  int FULL_DAY = 24*60;
+  Boolean[] freeStartTimes = new Boolean[FULL_DAY];
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    initializeFreeStartTimes();
+
+    for(Event e: events) {
+      if(!Collections.disjoint(e.getAttendees(), request.getAttendees())) {
+        busy(e, (int)request.getDuration());
+      }
+    }
+
+    return createFreeTimeRanges((int)request.getDuration());
+  }
+
+  private void initializeFreeStartTimes() {
+    Arrays.fill(freeStartTimes, Boolean.TRUE);
+  }
+
+  private void busy(Event e, int duration) {
+    for(int i = (e.getWhen().start()-duration); i < e.getWhen().end(); ++i) {
+      if(i < 0) {
+        i = 0;
+      }
+      if(i==(8*60+30)) {
+        System.out.println(e.getTitle());
+      }
+      freeStartTimes[i] = false;
+    }
+  }
+
+  private Collection<TimeRange> createFreeTimeRanges(int duration) {
+    Collection<TimeRange> freeTimeRanges = new ArrayList<TimeRange>();
+
+    if(duration > FULL_DAY) {
+      return freeTimeRanges;
+    }
+
+    boolean current = false;
+    int start = -1, end = -1;
+
+    for(int i = 0; i < FULL_DAY; ++i) {
+      if(freeStartTimes[i]) {
+        if(!current) {
+          start = i;
+          current = true;
+        }
+        end = i;
+      } else if(current == true) {
+        freeTimeRanges.add(TimeRange.fromStartEnd(start, end+duration, true));
+        current = false;
+      }
+    }
+    //printFreeStartTimes();
+    if(current == true) {
+      freeTimeRanges.add(TimeRange.fromStartEnd(start, end, true));
+    }
+    return freeTimeRanges;
+  }
+
+  public void printFreeStartTimes() {
```

Comment:
It would be better if this function could be more concise - i.e.
800-830: busy
830-900: free

Reply:
Also just used for debugging, removed in latest commit